### ISSUE TITLE
[LFX Pretest] Supoort for LSP4Ij

### ISSUE THREAD

**Issue by He1pa** [kcl-lang/intellij-kcl-lsp#4]
1. choose anyone plugin, provide research report(explain how it works, what we need to do, and a simple plan or anything else)
![image](https://github.com/user-attachments/assets/69303756-cdc8-4d43-bb81-22c1c39994c1)

2. try to build plugin locally(ensure that you understand the complete end-to-end process, so any plugin using LSP4IJ is ok)

**Comment by shishir-11** [comment]
Hey @He1pa can you please elaborate on the Pretest . Do we need to study other projects  that have already had support for LSP4IJ and prepare a report on that. By building a plugin locally do you mean we should try to recreate some of the plugin that are mentioned in the image or make one of our own .

**Comment by He1pa** [comment]
> > try to build plugin locally
> 
> to build this plugin [kcl-lang/intellij-kcl-lsp)](https://github.com/kcl-lang/intellij-kcl-lsp) ?
> 
> cc @Peefy

Any plugin using LSP4IJ

**Comment by He1pa** [comment]
> Hey @He1pa can you please elaborate on the Pretest . Do we need to study other projects that have already had support for LSP4IJ and prepare a report on that. By building a plugin locally do you mean we should try to recreate some of the plugin that are mentioned in the image or make one of our own .

Just learn how it works. In the report,   need to explain how it works,  what we need to do, and a simple plan. Local build is also to ensure that you understand the complete end-to-end process, so any plug-in using LSP4IJ is ok

**Comment by shishir-11** [comment]
@He1pa @Peefy 
Please take a look at this [doc](https://docs.google.com/document/d/1o-CjVr-f-tVwHPFeOUuhYtHLe-rCPphVlrLQwaRYcFQ/edit?usp=sharing) , I have just started drafting it and would love to receive your reviews on stuff that needs to be added or removed.

**Comment by He1pa** [comment]
> Hey @He1pa I was studying a bit about LSP and its usage and it was mentioned that any language server should be able to connect to any IDE given that the LSP specifications are similar across the different platforms. How different is it from vs code where kcl has an extension.
> 
> Can you also specify what features are expected to be delivered by this project as it would be convenient to understand them first. Thank you.

LSP was proposed by Microsoft, so it can be well supported in VSCode. But IntelliJ and IDEA do not belong to Microsoft. They only provide partial support for LSP, and it is not free. You can find the description in README.
This is also the reason why we tried to switch to LSP4IJ
We don't need to develop any features for LSP. We need to use LSP4IJ to integrate and start `kcl-language-server` just like vscode.

**Comment by He1pa** [comment]
> @He1pa @Peefy Please take a look at this [doc](https://docs.google.com/document/d/1o-CjVr-f-tVwHPFeOUuhYtHLe-rCPphVlrLQwaRYcFQ/edit?usp=sharing) , I have just started drafting it and would love to receive your reviews on stuff that needs to be added or removed.

We need more information about how this project uses LSP4IJ, such as which modules are used, how to set the local server path, how to establish a link with the server, etc.

**Comment by shishir-11** [comment]
I've tried reading the [docs](https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md#streamconnectionprovider-implementation) here and looked a bit into the implementation followed by the ZigBrains Plugins. 

I have added a little more to the [report](https://docs.google.com/document/d/1o-CjVr-f-tVwHPFeOUuhYtHLe-rCPphVlrLQwaRYcFQ/edit?usp=sharing) but had some doubts related to the location of the executable, so I would appreciate if you could take a look at it.

**Comment by He1pa** [comment]
> I've tried reading the [docs](https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md#streamconnectionprovider-implementation) here and looked a bit into the implementation followed by the ZigBrains Plugins.
> 
> I have added a little more to the [report](https://docs.google.com/document/d/1o-CjVr-f-tVwHPFeOUuhYtHLe-rCPphVlrLQwaRYcFQ/edit?usp=sharing) but had some doubts related to the location of the executable, so I would appreciate if you could take a look at it.

Sorry, I don't see what your doubts is. In vscode, we require users to add the server path to PATH in advance, so the startup command is `kcl-language-server`. you can see here: https://github.com/kcl-lang/vscode-kcl/blob/5c662f599a43fba60b61b2d728fedbd48f7dd76c/src/extension.ts#L64

I'm not sure，but according to the documentation, I guess we should set it to
```
public class MyLanguageServer extends ProcessStreamConnectionProvider {

    public MyLanguageServer() {
        List<String> commands = Arrays.asList("kcl-language-server");
        super.setCommands(commands);
    }
}
```

**Comment by shishir-11** [comment]
> Sorry, I don't see what your doubts is. In vscode, we require users to add the server path to PATH in advance, so the startup command is `kcl-language-server`. you can see here: https://github.com/kcl-lang/vscode-kcl/

Thanks for the clarification ,where can i get the code/binary for the server if it is already implemented. Or does this project expects a seperate server implementation too.

**Comment by He1pa** [comment]
> > Sorry, I don't see what your doubts is. In vscode, we require users to add the server path to PATH in advance, so the startup command is `kcl-language-server`. you can see here: https://github.com/kcl-lang/vscode-kcl/
> 
> Thanks for the clarification ,where can i get the code/binary for the server since it is already implemented for vs-code. Or does this project expects a seperate server implementation too if the same one won't work for intellij.

Server work for all client, now, we have vscode, nvim and intellij idea. you can find code from https://github.com/kcl-lang/kcl/tree/main/kclvm/tools/src/LSP  and download the binary from https://www.kcl-lang.io/docs/user_docs/getting-started/install#install-language-server

**Comment by shishir-11** [comment]
I've been trying to setup the plugin in intellij for some time and have consistently getting the error
```
Cannot resolve plugin 'com.redhat.devtools.lsp4ij' in dependencies (Reference: Plugin Dependencies) 
```
I have tried changing dependencies in the build.gradle.kts file and also reinstalled the lsp4ij plugins but have been getting the same thing. I have asked about it on the lsp4ij community but waiting for a response and would appreciate any help i receive.

**Comment by RehanChalana** [comment]
Hi @He1pa ,

I wanted to share that I am working on the pretest and have made some progress understanding the LSP and LSP4J. Currently, I am working on the report file and the local build of a plugin using LSP4J. I've found the learning process quite engaging.

Best regards

**Comment by RehanChalana** [comment]
> Research how this plugin uses LSP4IJ and a plan about what we need to do with kcl

I've researched how this plugin uses LSP4IJ, following the minimal configuration provided in the [docs](https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md#streamconnectionprovider-implementation).

I successfully applied similar minimal configurations for the kcl-lang with kcl-language-server, and the plugin worked well. However, syntax highlighting is not yet functioning.

![Screenshot 2024-08-05 224038](https://github.com/user-attachments/assets/6367f752-dc19-4a77-9bc9-40283aa44672)

Do I need to submit a report (along with plan for kcl-lang) on my implementations, given that the content might be limited, or should I submit a report(along with a plan for kcl-lang) on the implementations shown in the picture?


> 1. choose anyone plugin, provide research report(explain how it works, what we need to do, and a simple plan or anything else)

**Comment by RehanChalana** [comment]
sorry i was not able to convey my intentions in the previous comment

I've built the local implementation of plugin for the XML lang using LSP4iJ, following the minimal configuration provided in the [docs](https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md#streamconnectionprovider-implementation).

I successfully applied similar minimal configurations for the kcl-lang with kcl-language-server, and the plugin worked well. However, syntax highlighting is not yet functioning.

![Screenshot 2024-08-05 224038](https://github.com/user-attachments/assets/6367f752-dc19-4a77-9bc9-40283aa44672)

Do I need to research and submit a report (along with plan for kcl-lang) on my implementations, given that the content might be limited, or should I research and submit a report(along with a plan for kcl-lang) on the implementations shown in the picture?


> 1. choose anyone plugin, provide research report(explain how it works, what we need to do, and a simple plan or anything else)
